vim: Refactor and fix multiline operations

5brian opened the pull request:
Changes:
- [x] Cursor at the start during yank operations on objects (`yip`, `yab` etc). 
- [x] Refactors this: Trim all leading and trailing whitespace from inner multiline bracket selection.
  - This leaves a nicely indented line when doing `ci{` `vi{d` etc
  - [x] Checks for empty selection
  - [x] Removed moving cursor to the start in visual bracket operations

This cleans up the previous implementation by providing a simpler check in `surrounding_markers`, instead of calling a new function in `expand_object`. No functionality was changed there except for handling the empty selection and removing some cursor adjustments that should not have been there after further testing.

Release Notes:

- N/A

5brian committed:
check for empty range

5brian committed:
remove fn preserve indented newline, add non whitespace check

ConradIrwin commented:
Thank you! Blazingly fast as usual :D.

Could you please incorporate the original example from the issue as a test

5brian committed:
test whitespace selection

5brian commented:
@ConradIrwin I unfortunately have a midterm tomorrow that i need to finish studying for... Since theres still a bit more to this i don't think i will be able to finish this pr today, sorry

Im still trying to find out when neovim visual has the cursor at the start and when it has it at the end. From my minimal testing both brackets !around and around have it at the start. I just need to figure out where to flip it i think. 

If you need to finish the pr today please go ahead! if not i can hopefully get it done by the next few days

5brian committed:
fix visual cursor placement and fix test

5brian committed:
only reverse bracket objects

5brian committed:
update cursor in test

5brian committed:
tests

5brian committed:
update cursor, switch to vim test

5brian committed:
..

5brian committed:
update 1 test

5brian committed:
.

5brian commented:
Still some behaviors that do not match nvim yet:
- When yanking inside brackets like yib, cursor should move to the front.
- When in visual block, using text objects should stay in visual block after the operation.

5brian committed:
r

5brian committed:
last test fix

5brian committed:
..

ConradIrwin commented:
I'm going to fix the crash for now: https://github.com/zed-industries/zed/pull/25138, but sounds like we should make bigger changes here?

Happy to wait until you're done with exams

5brian commented:
Sounds good, thanks! will continue working on these

> When yanking inside brackets like yib, cursor should move to the front.
> When in visual block, using text objects should stay in visual block after the operation.

5brian committed:
yank: move cursor to start for some objects

5brian committed:
for objects, store start positions instead

5brian committed:
Merge branch 'main' into vim-expand-panic

5brian committed:
use updated yank in test

5brian committed:
find last non whitespace instead of last newline

5brian committed:
simplify?

5brian committed:
ignore quotes

5brian commented:
@ConradIrwin 

For `vi{` etc:
I just realized that since in nvim, whether if multiline or not, operations on any type of brackets do not have any lead/trailing whitespace selected, this means i was able to simplify the functionality of keeping the newline since it is the same as skipping all the whitespace.

For `vib`:
The problem is that `vib` does select everything including all the whitespace inside the brackets in nvim. From my understanding, there is no distinction between `b` and `(` in Zed's vim, so i'm not too sure how to proceed.

|`vi(`|`vib`|
|---|---|
|![image](https://github.com/user-attachments/assets/753ce616-ceca-467f-b58c-78024880120e)|![image](https://github.com/user-attachments/assets/9f6b00f7-f25d-4ff0-a2a3-8076a8457dda)|

5brian committed:
comment

ConradIrwin commented:
@5brian when I do `vi(` in `nvim --clean` it acts like your `vib` screenshot - i.e. in this case it does select all the whitespace. Do you have other plugins that are changing this behavior?

5brian commented:
Ah, just realized i had mini.nvim affecting the selections. Will redo this

5brian committed:
undo cursor fixup, only adjust selection on multiline

5brian committed:
undo tests

5brian committed:
cursor and mode

5brian committed:
persistant visual block

5brian committed:
switch test context for persistant vblock

5brian committed:
test and rm unnecessary

5brian committed:
only switch back to vblock or vline

5brian committed:
test persistant vline

5brian commented:
@ConradIrwin should be good for a review now, updated the description with the changes

Just confused about a test case i added for having visual line persist after using objects. 
Performing `shift-v a {` from inside a brace should result in the whole function including the declaration being selected, in visual line mode. I added the test with the set and assert exactly like that, when testing locally it works, but the ci test is failing -- i'm probably missing something but i cant think of anything

|assert|local build|ci|
|---|---|---|
|![image](https://github.com/user-attachments/assets/739f8aad-44dd-458e-a272-21c279954df2)|![image](https://github.com/user-attachments/assets/e7d759d0-8a6f-4791-bebc-426e845fb094)|![image](https://github.com/user-attachments/assets/11ee69a1-f6b8-414a-91ac-6f4e85491541)|

ConradIrwin commented:
The line mode selections are not reflected in the positions of the « »; even though the selection looks like it covers the whole line. If you hit `v` in that situation you can see where the real selection is, which should match what the test is assuming

5brian committed:
fix test

5brian committed:
Merge branch 'zed-industries:main' into vim-expand-panic

5brian commented:
Thank you

ConradIrwin commented:
@5brian can we keep using the NeovimBackedtestcontext for either of these tests given the above? Ideally we are exactly matching neovim when it has the features we have

5brian commented:
@ConradIrwin from my testing against nvim --clean

        cx.simulate_keystrokes("a ]");
        cx.assert_state("hello (in «[parens]ˇ» o)", Mode::VisualBlock);

Switches to visual mode instead of staying in visual block

        cx.simulate_keystrokes("shift-v a {");
        cx.assert_state(
            indoc! {
                "func empty(a string) bool «{
                   if a == \"\" {
                      return true
                   }
                   return false
                }ˇ»"
            },

Switches to visual mode instead of staying in visual line

I have them both staying in their original mode after the object operations, let me know if this is something i should change

ConradIrwin commented:
We should copy nvim here; if you want to select inside `[]`, you don't want to be in line mode anymore.

5brian committed:
revert test

5brian committed:
cleanup

5brian committed:
fix

5brian commented:
@ConradIrwin Just pushed the changes up, that makes more sense, not sure why that happened i think i got confused somewhere along the way. Sorry the pull request is a mess haha, hopefully everything looks good now! Updated the description as well

ConradIrwin commented:
Thanks!
